README.md
# ForwardProxy

![Gem Version][3] ![Gem][1] ![Build Status][2]

Minimal forward proxy using 150LOC and only standard libraries. Useful for development, testing, and learning.

```
$ forward-proxy --binding 0.0.0.0 --port 3182 --threads 2
I, [2021-07-04T10:33:32.947653 #1790]  INFO -- : Listening 0.0.0.0:3182
I, [2021-07-04T10:33:32.998298 #1790]  INFO -- : CONNECT raw.githubusercontent.com:443 HTTP/1.1
```

## Features

- Supports HTTPS `CONNECT` tunneling and HTTP `GET`, `POST`, and `HEAD` proxy request methods.
- Configurable client connection thread pool and timeout.
- `Transfer-Encoding: chunked` and response streaming.
- Automated tests.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'forward-proxy'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install forward-proxy

## Usage

### CLI

```
forward-proxy
```

```
Usage: forward-proxy [options]
    -p, --port=PORT                  Bind to specified port. Default: 9292
    -b, --binding=BINDING            Bind to the specified ip. Default: 127.0.0.1
    -t, --timeout=TIMEOUT            Specify the connection timeout in seconds. Default: 300
    -c, --threads=THREADS            Specify the number of client threads. Default: 128
    -h, --help                       Prints this help.
```

### Library

```ruby
require 'forward_proxy'

proxy = ForwardProxy::Server.new(
  bind_address: "127.0.0.1",
  bind_port: 3000
)

proxy.start
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/jamesmoriarty/forward-proxy. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/jamesmoriarty/forward-proxy/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the ForwardProxy project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/jamesmoriarty/forward-proxy/blob/master/CODE_OF_CONDUCT.md).

[1]: https://img.shields.io/gem/dt/forward-proxy
[2]: https://github.com/jamesmoriarty/forward-proxy/workflows/Continuous%20Integration/badge.svg?branch=main
[3]: https://img.shields.io/gem/v/forward-proxy


CHANGELOG.md
# CHANGELOG

## 0.9.1

- bugfix chunked length.

## 0.9.0

- bugfix chunked response behavior.

## 0.8.0

- add http `head` support.
- add http `501` not implemented status response.

## 0.7.0

- bugfix default threads from `4` to `128`.

## 0.6.0

- add connection timeout to stop tracking connection from saturating client threads.
- add cli flats for connection timeout `-t` and `--timeout`.
- change cli short flag `-t` to `-c` for `--threads`.

## 0.5.0

- increase default threads from `32` to `128`.

## 0.2.0

- Extract errors into module.

## 0.1.5

- stream http proxy requests.
- call shutdown on `Interrupt`
- remove waiting for the thread-pool to finish on shutdown and just close the server conn.

## 0.1.4

- `accept` connections via thread pool.
- catch and throw custom http parse error.
- `Server:` header added for default error 502 response.
- rescue and logs socket error `Errno::EBADF`.

## 0.1.3

- Refine CLI help text.

## 0.1.2

- Wrap ThreadPool in ForwardProxy module.

test/test_helper.rb
require 'openssl'
require 'webrick'
require 'webrick/https'

require "minitest/reporters"
Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new

require 'simplecov'
SimpleCov.start

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "forward_proxy"

require "minitest/autorun"

def root_key
  @root_key ||= begin
    OpenSSL::PKey::RSA.new 2048 # the CA's public/private key
  end
end

def root_ca
  @root_ca ||= begin
    OpenSSL::X509::Certificate.new.tap do |root_ca|
      root_ca.version = 2 # cf. RFC 5280 - to make it a "v3" certificate
      root_ca.serial = 1
      root_ca.subject = OpenSSL::X509::Name.parse "/DC=org/DC=ruby-lang/CN=Ruby CA"
      root_ca.issuer = root_ca.subject # root CA's are "self-signed"
      root_ca.public_key = root_key.public_key
      root_ca.not_before = Time.now
      root_ca.not_after = root_ca.not_before + 2 * 365 * 24 * 60 * 60 # 2 years validity
      ef = OpenSSL::X509::ExtensionFactory.new
      ef.subject_certificate = root_ca
      ef.issuer_certificate = root_ca
      root_ca.add_extension(ef.create_extension("basicConstraints", "CA:TRUE", true))
      root_ca.add_extension(ef.create_extension("keyUsage", "keyCertSign, cRLSign", true))
      root_ca.add_extension(ef.create_extension("subjectKeyIdentifier", "hash", false))
      root_ca.add_extension(ef.create_extension("authorityKeyIdentifier", "keyid:always", false))
      root_ca.sign(root_key, OpenSSL::Digest::SHA256.new)

      add_trusted_cert(root_ca)
    end
  end
end

def add_trusted_cert(cert)
  File.write("/tmp/Ruby-CA.crt", cert.to_s)

  case RUBY_PLATFORM
  when /linux/
    `sudo cp /tmp/Ruby-CA.crt /usr/local/share/ca-certificates/Ruby-CA.crt`
    `sudo update-ca-certificates`
  when /darwin/
    `sudo security delete-certificate -c 'Ruby CA'`
    `sudo security add-trusted-cert -d -r trustRoot -k "/Library/Keychains/System.keychain" /tmp/Ruby-CA.crt`
  else
    throw "#{RUBY_PLATFORM}: unexpected platform"
  end
end

def key
  @key ||= begin
    OpenSSL::PKey::RSA.new 2048
  end
end

def cert(domains: ["IP:127.0.0.1", "DNS:localhost", "DNS:#{WEBrick::Utils::getservername()}"])
  @cert ||= begin
    OpenSSL::X509::Certificate.new.tap do |cert|
      cert.version = 2
      cert.serial = 2
      cert.subject = OpenSSL::X509::Name.parse "/DC=org/DC=ruby-lang/CN=Ruby Certificate"
      cert.issuer = root_ca.subject # root CA is the issuer
      cert.public_key = key.public_key
      cert.not_before = Time.now
      cert.not_after = cert.not_before + 1 * 365 * 24 * 60 * 60 # 1 years validity
      ef = OpenSSL::X509::ExtensionFactory.new
      ef.subject_certificate = cert
      ef.issuer_certificate = root_ca
      cert.add_extension(ef.create_extension("keyUsage", "digitalSignature", true))
      cert.add_extension(ef.create_extension("subjectKeyIdentifier", "hash", false))
      cert.add_extension(ef.create_extension("subjectAltName",
                                             domains.join(" ,"), false))
      cert.sign(root_key, OpenSSL::Digest::SHA256.new)
    end
  end
end

def with_dest(uri, middleware = {})
  options = {
    BindAddress: uri.host,
    Port: uri.port,
    DocumentRoot: Dir.pwd
  }

  options.merge!(
    SSLEnable: true,
    SSLCertificate: cert,
    SSLPrivateKey: key,
    SSLCertName: [["CN", WEBrick::Utils::getservername]]
  ) if uri.scheme == 'https'

  server = WEBrick::HTTPServer.new(options)

  middleware.each { |path, app| server.mount_proc(path, app) }

  server_thread = Thread.new { server.start }

  sleep 0.050 # 50ms

  yield if block_given?

  server.shutdown

  server_thread.join
end

def with_proxy(uri, bind_address: "127.0.0.1", bind_port: 3000, timeout: 1, threads: 32, logger: Logger.new(STDOUT,
                                                                                                            level: :error))
  proxy = ForwardProxy::Server.new(
    bind_address: bind_address,
    bind_port: bind_port,
    logger: logger,
    timeout: timeout
  )

  proxy_thread = Thread.new { proxy.start }

  sleep 0.050 # 50ms

  Net::HTTP.start(
    uri.hostname,
    uri.port,
    proxy.bind_address,
    proxy.bind_port,
    use_ssl: uri.scheme == 'https'
  ) do |http|
    yield http if block_given?
  end

  proxy.shutdown

  proxy_thread.join
end


test/forward_proxy_test.rb
require "test_helper"

class ForwardProxyTest < Minitest::Test
  def test_handle_get
    with_dest(uri = URI('http://127.0.0.1:8000/test/index.txt')) do
      with_proxy(uri) do |http|
        resp = http.request Net::HTTP::Get.new(uri)

        assert_equal "200", resp.code
        assert_match "WEBrick", resp['server']
        assert_equal "HTTP/1.1 ForwardProxy", resp['via']
        assert_equal "hello world", resp.body
      end
    end
  end

  def test_handle_head
    with_dest(uri = URI('http://127.0.0.1:8000/test/index.txt')) do
      with_proxy(uri) do |http|
        resp = http.request Net::HTTP::Head.new(uri)

        assert_equal "200", resp.code
        assert_match "WEBrick", resp['server']
        assert_equal "HTTP/1.1 ForwardProxy", resp['via']
        assert_equal nil, resp.body
      end
    end
  end

  def test_handle_post
    with_dest(uri = URI('http://127.0.0.1:8000/test/index.txt')) do
      with_proxy(uri) do |http|
        resp = http.request Net::HTTP::Post.new(uri)

        assert_equal "405", resp.code
        assert_match "WEBrick", resp['server']
        assert_equal "HTTP/1.1 ForwardProxy", resp['via']
        assert_match "Method Not Allowed", resp.body
      end
    end
  end

  def test_handle_with_unsupported_method
    with_dest(uri = URI('http://127.0.0.1:8000/test/index.txt')) do
      with_proxy(uri) do |http|
        resp = http.request Net::HTTP::Trace.new(uri)

        assert_equal "501", resp.code
        assert_equal "HTTP/1.1 ForwardProxy", resp['via']
        assert_equal "", resp.body
      end
    end
  end

  def test_handle_tunnel
    with_dest(uri = URI('https://127.0.0.1:8000/test/index.txt')) do
      with_proxy(uri) do |http|
        resp = http.request Net::HTTP::Get.new(uri)

        assert_equal "200", resp.code
        assert_match "WEBrick", resp['server']
        assert_equal "hello world", resp.body
      end
    end
  end

  def test_handle_error_with_timeout
    app = Proc.new do |req, res|
      rd, wr = IO.pipe
      res.body = rd
      sleep 1 # 1000ms
      wr.close
    end

    with_dest(uri = URI('http://127.0.0.1:8000/stream'), { '/stream' => app }) do
      with_proxy(uri, timeout: 1/20r) do |http| # 50ms
        resp = http.request Net::HTTP::Get.new(uri)

        assert_equal "504", resp.code
      end
    end
  end

  def test_stream_response_body
    called = 0
    content_body = "0123456789" * 1_000
    response_body = ""

    app = Proc.new do |req, res|
      res.status = 200
      res['Content-Type'] = 'text/plain'
      res.chunked = true
      res.body = content_body
      res['Content-Length'] = content_body.length
    end

    with_dest(uri = URI('http://127.0.0.1:8000/chunked'), { '/chunked' => app }) do
      with_proxy(uri) do |http|
        begin
          http.request Net::HTTP::Get.new(uri) do |response|
            response.read_body do |chunk|
              response_body += chunk
              called += 1
            end
          end
        rescue EOFError # https://bugs.ruby-lang.org/issues/14972
          called += 1
        end

        assert 1 < called
        assert_match content_body, response_body
      end
    end
  end

  def test_logger
    with_dest(uri = URI('http://127.0.0.1:8000/test/index.txt')) do
      with_proxy(uri, logger: Logger.new(io = StringIO.new)) do |http|
        resp = http.request Net::HTTP::Get.new(uri)

        assert_equal "200", resp.code
        assert_match uri.path, io.string
      end
    end
  end
end


lib/forward_proxy/server.rb
require 'logger'
require 'socket'
require 'timeout'
require 'net/http'
require 'webrick'
require 'forward_proxy/errors/connection_timeout_error'
require 'forward_proxy/errors/http_method_not_implemented'
require 'forward_proxy/errors/http_parse_error'
require 'forward_proxy/thread_pool'

module ForwardProxy
  class Server
    attr_reader :bind_address, :bind_port, :logger, :timeout

    def initialize(bind_address: "127.0.0.1", bind_port: 9292, threads: 128, timeout: 300, logger: default_logger)
      @bind_address = bind_address
      @bind_port = bind_port
      @logger = logger
      @thread_pool = ThreadPool.new(threads)
      @timeout = timeout
    end

    def start
      thread_pool.start

      @socket = TCPServer.new(bind_address, bind_port)

      logger.info("Listening #{bind_address}:#{bind_port}")

      loop do
        thread_pool.schedule(socket.accept) do |client_conn|
          begin
            Timeout::timeout(timeout, Errors::ConnectionTimeoutError, "connection exceeded #{timeout} seconds") do
              req = parse_req(client_conn)

              logger.info(req.request_line.strip)

              case req.request_method
              when METHOD_CONNECT then handle_tunnel(client_conn, req)
              when METHOD_GET, METHOD_HEAD, METHOD_POST then handle(client_conn, req)
              else
                raise Errors::HTTPMethodNotImplemented, "unsupported http method #{req.request_method}"
              end
            end
          rescue => e
            handle_error(client_conn, e)
          ensure
            client_conn.close
          end
        end
      end
    rescue Interrupt
      shutdown
    rescue IOError, Errno::EBADF => e
      logger.error(e.message)
    end

    def shutdown
      socket.close if socket
    end

    private

    attr_reader :socket, :thread_pool

    METHOD_CONNECT, METHOD_GET, METHOD_HEAD, METHOD_POST = "CONNECT", "GET", "HEAD", "POST"

    # The following comments are from the IETF document
    # "Hypertext Transfer Protocol -- HTTP/1.1: Basic Rules"
    # https://datatracker.ietf.org/doc/html/rfc2616#section-2.2

    # HTTP/1.1 defines the sequence CR LF as the end-of-line marker for all
    # protocol elements except the entity-body.
    HTTP_EOP = "\r\n"

    # The following comments are from the IETF document
    # "Hypertext Transfer Protocol (HTTP/1.1): Semantics and Content"
    # https://tools.ietf.org/html/rfc7231#section-4.3.6

    # A proxy MUST send an appropriate Via header field, as described
    # below, in each message that it forwards.  An HTTP-to-HTTP gateway
    # MUST send an appropriate Via header field in each inbound request
    # message and MAY send a Via header field in forwarded response
    # messages.
    HEADER_VIA = "HTTP/1.1 ForwardProxy"

    def handle_tunnel(client_conn, req)
      # The following comments are from the IETF document
      # "Hypertext Transfer Protocol (HTTP/1.1): Semantics and Content"
      # https://tools.ietf.org/html/rfc7231#section-4.3.6

      # A client sending a CONNECT request MUST send the authority form of
      # request-target (Section 5.3 of [RFC7230]); i.e., the request-target
      # consists of only the host name and port number of the tunnel
      # destination, separated by a colon.  For example,
      #
      #   CONNECT server.example.com:80 HTTP/1.1
      #   Host: server.example.com:80
      #
      host, port = *req['Host'].split(':')
      dest_conn = TCPSocket.new(host, port)

      # Any 2xx (Successful) response indicates that the sender (and all
      # inbound proxies) will switch to tunnel mode immediately after the
      # blank line that concludes the successful response's header section;
      # data received after that blank line is from the server identified by
      # the request-target.
      client_conn.write <<~eos.chomp
        HTTP/1.1 200 OK
        #{HTTP_EOP}
      eos

      # The CONNECT method requests that the recipient establish a tunnel to
      # the destination origin server identified by the request-target and,
      # if successful, thereafter restrict its behavior to blind forwarding
      # of packets, in both directions, until the tunnel is closed.
      [
        Thread.new { transfer(client_conn, dest_conn) },
        Thread.new { transfer(dest_conn, client_conn) }
      ].each(&:join)
    ensure
      dest_conn.close
    end

    def handle(client_conn, req)
      Net::HTTP.start(req.host, req.port) do |http|
        http.request(map_webrick_to_net_http_req(req)) do |resp|
          # The following comments are from the IETF document
          # "Hypertext Transfer Protocol (HTTP/1.1): Semantics and Content"
          # https://tools.ietf.org/html/rfc7231#section-4.3.6

          # An intermediary MAY combine an ordered subsequence of Via header
          # field entries into a single such entry if the entries have identical
          # received-protocol values.  For example,
          #
          #   Via: 1.0 ricky, 1.1 ethel, 1.1 fred, 1.0 lucy
          #
          # could be collapsed to
          #
          #   Via: 1.0 ricky, 1.1 mertz, 1.0 lucy
          #
          # A sender SHOULD NOT combine multiple entries unless they are all
          # under the same organizational control and the hosts have already been
          # replaced by pseudonyms.  A sender MUST NOT combine entries that have
          # different received-protocol values.
          headers = resp.to_hash.merge(Via: [HEADER_VIA, resp['Via']].compact.join(', '))

          client_conn.puts <<~eos.chomp
            HTTP/1.1 #{resp.code}
            #{headers.map { |header, value| "#{header}: #{value}" }.join("\n")}
            #{HTTP_EOP}
          eos

          # The following comments are taken from:
          # https://docs.ruby-lang.org/en/2.0.0/Net/HTTP.html#class-Net::HTTP-label-Streaming+Response+Bodies

          # By default Net::HTTP reads an entire response into memory. If you are
          # handling large files or wish to implement a progress bar you can
          # instead stream the body directly to an IO.
          resp.read_body do |chunk|
            # The following comments are taken from:
            # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding#directives

            # Data is sent in a series of chunks. The Content-Length header is omitted in this case and
            # at the beginning of each chunk you need to add the length of the current chunk in
            # hexadecimal format, followed by '\r\n' and then the chunk itself, followed by another
            # '\r\n'. The terminating chunk is a regular chunk, with the exception that its length
            # is zero. It is followed by the trailer, which consists of a (possibly empty) sequence of
            # header fields.
            client_conn << chunk.bytesize.to_s(16) + HTTP_EOP if resp['Transfer-Encoding'] == 'chunked'

            client_conn << chunk
          end
        end
      end
    end

    def handle_error(client_conn, err)
      status_code = case err
                    when Errors::ConnectionTimeoutError then 504
                    when Errors::HTTPMethodNotImplemented then 501
                    else
                      502
                    end

      client_conn.puts <<~eos.chomp
        HTTP/1.1 #{status_code}
        Via: #{HEADER_VIA}
        #{HTTP_EOP}
      eos

      logger.error(err.message)
      logger.debug(err.backtrace.join("\n"))
    end

    def default_logger
      Logger.new(STDOUT, level: :info)
    end

    def map_webrick_to_net_http_req(req)
      req_headers = Hash[req.header.map { |k, v| [k, v.first] }]

      klass = case req.request_method
              when METHOD_GET then Net::HTTP::Get
              when METHOD_HEAD then Net::HTTP::Head
              when METHOD_POST then Net::HTTP::Post
              else
                raise Errors::HTTPMethodNotImplemented
              end

      klass.new(req.path, req_headers)
    end

    def transfer(src_conn, dest_conn)
      IO.copy_stream(src_conn, dest_conn)
    rescue => e
      logger.warn(e.message)
    end

    def parse_req(client_conn)
      WEBrick::HTTPRequest.new(WEBrick::Config::HTTP).tap do |req|
        req.parse(client_conn)
      end
    rescue => e
      throw Errors::HTTPParseError.new(e.message)
    end
  end
end
